File: Services/ChannelResolver.swift
import Foundation

/// 入力URLから解決したチャンネル識別子。
enum ChannelIdentifier: Equatable {
    case channelId(String)   // UCxxxx
    case handle(String)      // @ を除いたハンドル
    case username(String)    // 旧 /user/ 形式
    case customName(String)  // /c/name または /name のカスタムURL
    case video(String)       // 動画URL（videoId）。channelId は API で解決する
}

/// YouTube チャンネルURL（または handle / channelId）を解析する。
/// ネットワークアクセスを行わない純粋関数なのでテストしやすい。
enum ChannelResolver {

    static func parse(_ rawInput: String) throws -> ChannelIdentifier {
        let input = rawInput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !input.isEmpty else { throw YouTubeAPIError.invalidChannelURL }

        // 1) 先頭が @ のハンドル単体
        if input.hasPrefix("@") {
            return try makeHandle(String(input.dropFirst()))
        }

        // 2) channelId 単体（UC + 22文字）
        if isChannelId(input) {
            return .channelId(input)
        }

        // 3) URLとして解析（スキーム省略も許容）
        let normalized = input.contains("://") ? input : "https://\(input)"
        guard let comps = URLComponents(string: normalized),
              let host = comps.host?.lowercased(),
              host.contains("youtube.com") || host.contains("youtu.be") else {
            throw YouTubeAPIError.invalidChannelURL
        }

        let segments = comps.path.split(separator: "/").map(String.init)

        // 短縮URL（youtu.be/VIDEOID）は常に動画。
        if host == "youtu.be" || host.hasSuffix(".youtu.be") {
            guard let id = segments.first, isVideoId(id) else {
                throw YouTubeAPIError.invalidVideoURL
            }
            return .video(id)
        }

        guard let first = segments.first else {
            throw YouTubeAPIError.invalidChannelURL
        }

        switch first.lowercased() {
        case "watch":
            // watch?v=VIDEOID（クエリで動画IDを渡す形式）
            guard let v = comps.queryItems?.first(where: { $0.name == "v" })?.value,
                  isVideoId(v) else {
                throw YouTubeAPIError.invalidVideoURL
            }
            return .video(v)
        case "shorts", "live", "embed", "v":
            guard segments.count >= 2, isVideoId(segments[1]) else {
                throw YouTubeAPIError.invalidVideoURL
            }
            return .video(segments[1])
        case "channel":
            guard segments.count >= 2, isChannelId(segments[1]) else {
                throw YouTubeAPIError.invalidChannelURL
            }
            return .channelId(segments[1])
        case "user":
            guard segments.count >= 2 else { throw YouTubeAPIError.invalidChannelURL }
            return .username(segments[1])
        case "c":
            guard segments.count >= 2 else { throw YouTubeAPIError.invalidChannelURL }
            return .customName(segments[1])
        default:
            if first.hasPrefix("@") {
                return try makeHandle(String(first.dropFirst()))
            }
            // youtube.com/SomeName のような旧カスタムURL
            return .customName(first)
        }
    }

    /// UC で始まる24文字の channelId かどうか。
    static func isChannelId(_ s: String) -> Bool {
        guard s.hasPrefix("UC"), s.count == 24 else { return false }
        return s.allSatisfy { $0.isLetter || $0.isNumber || $0 == "_" || $0 == "-" }
    }

    /// YouTube の videoId（11文字の英数字 + `_` `-`）かどうか。
    static func isVideoId(_ s: String) -> Bool {
        guard s.count == 11 else { return false }
        return s.allSatisfy { $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "_" || $0 == "-") }
    }

    /// 動画URL（または videoId 単体）から videoId を取り出す。動画URLでなければ nil。
    /// 共有された URL の種類を判別したい場面（テスト・ログ）のための補助 API。
    static func extractVideoId(from rawInput: String) -> String? {
        let input = rawInput.trimmingCharacters(in: .whitespacesAndNewlines)
        if isVideoId(input), !input.contains("/"), !input.contains(".") {
            return input
        }
        guard let identifier = try? parse(input), case .video(let id) = identifier else {
            return nil
        }
        return id
    }

    private static func makeHandle(_ h: String) throws -> ChannelIdentifier {
        let clean = h.trimmingCharacters(in: .whitespaces)
        guard !clean.isEmpty else { throw YouTubeAPIError.invalidChannelURL }
        return .handle(clean)
    }
}

File: android/app/src/main/java/com/deskflowlabs/channeltimelineviewer/network/ChannelResolver.kt
package com.deskflowlabs.channeltimelineviewer.network

import android.net.Uri

/**
 * 入力URLから解決したチャンネル識別子。
 *
 * iOS 版 `Services/ChannelResolver.swift` の移植。挙動を変えないこと
 * （どちらのアプリでも同じURLが同じように開ける必要がある）。
 */
sealed interface ChannelIdentifier {
    /** UCxxxx 形式のチャンネルID。 */
    data class ChannelId(val value: String) : ChannelIdentifier

    /** @ を除いたハンドル。 */
    data class Handle(val value: String) : ChannelIdentifier

    /** 旧 /user/ 形式。 */
    data class Username(val value: String) : ChannelIdentifier

    /** /c/name または /name のカスタムURL。 */
    data class CustomName(val value: String) : ChannelIdentifier

    /** 動画URL（videoId）。チャンネルIDは API で解決する。 */
    data class Video(val videoId: String) : ChannelIdentifier
}

/**
 * YouTube チャンネルURL（または handle / channelId）を解析する。
 * ネットワークアクセスをしない純粋関数なのでテストしやすい。
 */
object ChannelResolver {

    @Throws(YouTubeApiException::class)
    fun parse(rawInput: String): ChannelIdentifier {
        val input = rawInput.trim()
        if (input.isEmpty()) throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)

        // 1) 先頭が @ のハンドル単体
        if (input.startsWith("@")) return makeHandle(input.drop(1))

        // 2) channelId 単体（UC + 22文字）
        if (isChannelId(input)) return ChannelIdentifier.ChannelId(input)

        // 3) URLとして解析（スキーム省略も許容）
        val normalized = if (input.contains("://")) input else "https://$input"
        val uri = runCatching { Uri.parse(normalized) }.getOrNull()
        val host = uri?.host?.lowercase()
        if (host == null || !(host.contains("youtube.com") || host.contains("youtu.be"))) {
            throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)
        }

        val segments = (uri.path ?: "").split("/").filter { it.isNotEmpty() }

        // 短縮URL（youtu.be/VIDEOID）は常に動画。
        if (host == "youtu.be" || host.endsWith(".youtu.be")) {
            val id = segments.firstOrNull()
            if (id == null || !isVideoId(id)) throw YouTubeApiException(YouTubeApiError.InvalidVideoUrl)
            return ChannelIdentifier.Video(id)
        }

        val first = segments.firstOrNull() ?: throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)

        return when (first.lowercase()) {
            "watch" -> {
                val v = runCatching { uri.getQueryParameter("v") }.getOrNull()
                if (v == null || !isVideoId(v)) throw YouTubeApiException(YouTubeApiError.InvalidVideoUrl)
                ChannelIdentifier.Video(v)
            }

            "shorts", "live", "embed", "v" -> {
                val id = segments.getOrNull(1)
                if (id == null || !isVideoId(id)) throw YouTubeApiException(YouTubeApiError.InvalidVideoUrl)
                ChannelIdentifier.Video(id)
            }

            "channel" -> {
                val id = segments.getOrNull(1)
                if (id == null || !isChannelId(id)) throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)
                ChannelIdentifier.ChannelId(id)
            }

            "user" -> {
                val name = segments.getOrNull(1) ?: throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)
                ChannelIdentifier.Username(name)
            }

            "c" -> {
                val name = segments.getOrNull(1) ?: throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)
                ChannelIdentifier.CustomName(name)
            }

            else ->
                if (first.startsWith("@")) makeHandle(first.drop(1))
                // youtube.com/SomeName のような旧カスタムURL
                else ChannelIdentifier.CustomName(first)
        }
    }

    /** UC で始まる24文字の channelId かどうか。 */
    fun isChannelId(value: String): Boolean =
        value.startsWith("UC") &&
            value.length == 24 &&
            value.all { it.isLetterOrDigit() || it == '_' || it == '-' }

    /** YouTube の videoId（11文字の英数字 + `_` `-`）かどうか。 */
    fun isVideoId(value: String): Boolean =
        value.length == 11 &&
            value.all { it.code < 128 && (it.isLetterOrDigit() || it == '_' || it == '-') }

    /**
     * 動画URL（または videoId 単体）から videoId を取り出す。動画URLでなければ null。
     */
    fun extractVideoId(rawInput: String): String? {
        val input = rawInput.trim()
        if (isVideoId(input) && !input.contains("/") && !input.contains(".")) return input
        val identifier = runCatching { parse(input) }.getOrNull()
        return (identifier as? ChannelIdentifier.Video)?.videoId
    }

    private fun makeHandle(handle: String): ChannelIdentifier {
        val clean = handle.trim()
        if (clean.isEmpty()) throw YouTubeApiException(YouTubeApiError.InvalidChannelUrl)
        return ChannelIdentifier.Handle(clean)
    }
}
